### PR TITLE
Parallel reading from replicas over Distributed

### PR DESCRIPTION
**Pull request by devcrafter**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Enable parallel reading from replicas over distributed table. Related to https://github.com/ClickHouse/ClickHouse/issues/49708

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
This is an automated comment for commit e91b97248695ccd2ac2ac10b15f7a450dc5e14ff with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/53005/e91b97248695ccd2ac2ac10b15f7a450dc5e14ff/ci_running.html)
The overall status of the commit is 🔴 failure

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🔴 failure</td></tr>
</table>

File: src/Processors/QueryPlan/ReadFromRemote.cpp [hunk 1]
```diff
@@ -240,7 +240,8 @@ void ReadFromRemote::addPipe(Pipes & pipes, const ClusterProxy::SelectStreamFact
             shard.shard_info.pool, query_string, output_stream->header, context, throttler, scalars, external_tables, stage);
 
     remote_query_executor->setLogger(log);
-    remote_query_executor->setPoolMode(PoolMode::GET_MANY);
+    if (context->getParallelReplicasMode() == Context::ParallelReplicasMode::READ_TASKS)
```
Comment (alexey-milovidov) [inline src/Processors/QueryPlan/ReadFromRemote.cpp:243]: A comment is needed.

**Comment by alexey-milovidov** [comment]
> Propagate setting cluster_for_parallel_replicas to shards

Just in case, the default value for cluster_for_parallel_replicas for queries over distributed tables should be set to the cluster of the distributed table.

**Comment by devcrafter** [comment]
> > Propagate setting cluster_for_parallel_replicas to shards
> 
> Just in case, the default value for cluster_for_parallel_replicas for queries over distributed tables should be set to the cluster of the distributed table.

Exactly this is done in the commit, - distributed table cluster is passed to shards as `_cluster_for_parallel_replicas` scalar value. And the assumption that all nodes have the same definition of the cluster.

**Review by nikitamikhaylov — APPROVED** [review]

**Comment by devcrafter** [comment]
Upgrade check failure - see https://github.com/ClickHouse/ClickHouse/pull/53611

**Comment by alexey-milovidov** [comment]
Just in case - does it fix the green progress bar?

**Comment by devcrafter** [comment]
> Just in case - does it fix the green progress bar?

I didn't check it, but don't think so. Will address it separately.

File: src/Storages/StorageReplicatedMergeTree.cpp [hunk 2]
```diff
@@ -5157,7 +5157,9 @@ void StorageReplicatedMergeTree::readParallelReplicasImpl(
 {
     auto table_id = getStorageID();
 
-    auto parallel_replicas_cluster = local_context->getCluster(local_context->getSettingsRef().cluster_for_parallel_replicas);
+    auto scalars = local_context->hasQueryContext() ? local_context->getQueryContext()->getScalars() : Scalars{};
```
Comment (antonio2368) [inline src/Storages/StorageReplicatedMergeTree.cpp:5160]: is there a reason we get scalars here?

Comment (devcrafter) [reply src/Storages/StorageReplicatedMergeTree.cpp:5160]: It's leftovers. Will remove it. Thanks

Comment (devcrafter) [reply src/Storages/StorageReplicatedMergeTree.cpp:5160]: @rschu1ze Shouldn't it be caught during our build? (with `-Wunused-variable`)

Comment (rschu1ze) [reply src/Storages/StorageReplicatedMergeTree.cpp:5160]: It should and it does. Unfortunately, the compiler only complains about simple cases like `auto scalars = Scalars{};`.  Whenever functions with potential side effects (like `hasQueryContext()` or `getScalars()`) are involved, the warning doesn't trigger.

**Pull request merged**